File: DemoApp Objective-C/CASSample/SupportFiles/Main/MainVC.m
//
//  MainVC.m
//  CASSample
//

#import "MainVC.h"
#import "MainTVCell.h"
#import "FormatsSection.h"
#import "AppCoordinator.h"

@interface MainVC () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) IBOutlet UITableView *tableView;
@property (nonatomic, strong) NSArray<NSNumber *> *formats;

@end

@implementation MainVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"Select Ad format";
    self.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeAlways;
    
    self.formats = [FormatsSectionHelper allCases];
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    [MainTVCell registerForTableView:self.tableView];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.formats.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 24.0;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewAutomaticDimension;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MainTVCell *cell = [tableView dequeueReusableCellWithIdentifier:[MainTVCell cellIdentifier] forIndexPath:indexPath];
    FormatsSection format = [self.formats[indexPath.row] integerValue];
    NSString *title = [FormatsSectionHelper titleForFormat:format];
    [cell configureCellWithTitle:title];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    FormatsSection selectedFormat = [self.formats[indexPath.row] integerValue];
    
    switch (selectedFormat) {
        case FormatsSectionAppOpen:
            [self.coordinator navigateToAppOpenAd];
            break;
        case FormatsSectionBanner:
            [self.coordinator navigateToBannerAd];
            break;
        case FormatsSectionNative:
            [self.coordinator navigateToNativeAd];
            break;
        case FormatsSectionNativeTemplate:
            [self.coordinator navigateToNativeTemplateAd];
            break;
        case FormatsSectionInterstitial:
            [self.coordinator navigateToInterstitialAd];
            break;
        case FormatsSectionRewarded:
            [self.coordinator navigateToRewardedAd];
            break;
    }
}

@end

File: DemoApp Swift/CASSample/SupportFiles/Main/MainVC.swift
//
//  MainVC.swift
//  CASSample
//
//  Copyright © 2025 Clever Ads Solutions. All rights reserved.
//

import UIKit

class MainVC: UIViewController {

    @IBOutlet private weak var tableView: UITableView!

    weak var coordinator: AppCoordinator?
    private var formats: [FormatsSection] = FormatsSection.allCases
    
    override func viewDidLoad() {
        super.viewDidLoad()
        //
        navigationItem.largeTitleDisplayMode = .always
        navigationItem.title = "Select Ad format"
        setupTableView()
    }
    
    private func setupTableView() {
        tableView.delegate = self
        tableView.dataSource = self
        //
        MainTVCell.registerForTableView(tableView)
    }
}

extension MainVC: UITableViewDelegate, UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return formats.count
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 24.0
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableView.automaticDimension
    }
        
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: MainTVCell.cellIdentifier, for: indexPath) as? MainTVCell else { return UITableViewCell() }
        cell.configureCell(formats[indexPath.row].title)
        return cell
    }
        
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let selectedItem = formats[indexPath.row]
        
        switch selectedItem {
        case .appOpen:
            coordinator?.navigateToAppOpenAd()
        case .banner:
            coordinator?.navigateToBannerAd()
        case .native:
            coordinator?.navigateToNativeAd()
        case .nativeTemplate:
            coordinator?.navigateToNativeTemplate()
        case .interstitial:
            coordinator?.navigateToInterstitialAd()
        case .rewarded:
            coordinator?.navigateToRewardedAd()
        }
    }
}
